### PR TITLE
[core] GCS version auth [5/n]

### PR DESCRIPTION
**Pull request by vitsai**
Add AuthType template parameter to authenticate against cluster token.

(Moved authentication logic over from #35014 to make the PR smaller.)

## Related issue number

#34763

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by vitsai** [comment]
cc: @bveeramani release-blocking PR

File: src/ray/gcs/gcs_client/accessor.h [hunk 1]
```diff
@@ -269,7 +269,7 @@ class JobInfoAccessor {
 class NodeInfoAccessor {
  public:
   NodeInfoAccessor() = default;
-  explicit NodeInfoAccessor(GcsClient *client_impl);
+  NodeInfoAccessor(GcsClient *client_impl);
```
Comment (scv119) [inline src/ray/gcs/gcs_client/accessor.h:272]: huh, why remove explicit?

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 2]
```diff
@@ -81,9 +85,11 @@ class RAY_EXPORT GcsClient : public std::enable_shared_from_this<GcsClient> {
   /// Connect to GCS Service. Non-thread safe.
   /// This function must be called before calling other functions.
   /// \param instrumented_io_context IO execution service.
+  /// \param cluster_id Optional cluster ID to provide to the client.
```
Comment (scv119) [inline src/ray/gcs/gcs_client/gcs_client.h:88]: should we add comments somewhere on expected behavior of setting cluster_id to Nil?

File: src/ray/raylet/main.cc [hunk 3]
```diff
@@ -170,153 +170,156 @@ int main(int argc, char *argv[]) {
   RAY_CHECK_OK(gcs_client->Connect(main_service));
   std::unique_ptr<ray::raylet::Raylet> raylet;
 
-  RAY_CHECK_OK(gcs_client->Nodes().AsyncGetInternalConfig(
-      [&](::ray::Status status,
-          const boost::optional<std::string> &stored_raylet_config) {
-        RAY_CHECK_OK(status);
-        RAY_CHECK(stored_raylet_config.has_value());
-        RayConfig::instance().initialize(stored_raylet_config.get());
+  auto f = std::async(std::launch::async, [&]() {
```
Comment (scv119) [inline src/ray/raylet/main.cc:173]: any reason making this change?

File: src/ray/rpc/client_call.h [hunk 4]
```diff
@@ -227,7 +244,7 @@ class ClientCallManager {
   /// -1 means it will use the default timeout configured for the handler.
   ///
   /// \return A `ClientCall` representing the request that was just sent.
-  template <class GrpcService, class Request, class Reply>
+  template <class GrpcService, class Request, class Reply, bool Insecure>
```
Comment (scv119) [inline src/ray/rpc/client_call.h:247]: super nit: Insecure -> VerifyClusterId

**Review by scv119 — CHANGES_REQUESTED** [review]
also the merge conflict need to be fixed.

**Pull request closed without merging**